design/baremetal-operator/early-cleaning.md

```diff
@@ -0,0 +1,126 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# resetting BMC through Ironic prior to BMH deployment
+
+This document proposes running driver-specific out-of-band clean
+steps.
+
+## Status
+
+implementable
+
+## Summary
+
+Depending on the selected driver, BMO will run zero or more out-of-band clean
+steps to ensure the node and the BMC are in the right state. Out-of-band clean
+steps are executed via BMC (e.g. using Redfish protocol) without booting the
+ironic agent (IPA).
+
+The first drivers to receive such handling will be `idrac` and
+`idrac-virtualmedia`. For them we will run the `management.known_good_state`
+clean step that resets iDRAC and clears its job queue.
+
+It is important that these actions run before we try to boot anything on the
+target node, be it the inspection ramdisk or a 3rd party installer ISO.
+
+## Motivation
+
+iDRAC will refuse to create a job if there is already a pending job of the same
+type. To make sure we start in a coherent state, we will reset iDRAC before
+doing anything else to it.
+
+### Goals
+
+* Provide a mechanism to run out-of-band clean steps before inspection.
+* Run `known_good_state` clean step for iDRAC.
+
+### Non-Goals
+
+* Run any in-band (i.e. using a ramdisk) clean steps.
+* Allow a user to provide the steps to run.
+
+## Proposal
+
+## Design Details
+
+All clean steps run asynchronously in the ironic API, even out-of-band steps
+that are reasonable fast (an order of seconds). When entering the `Registering`
+state, we will check if the driver supplies any reset clean steps. If yes, we
+will start manual cleaning and wait. Otherwise we'll move on.
```

Comment:
The problem here is knowing whether you've done the reset, since the Ironic Node looks the same both before and after - it's 'manageable' in both cases (it would be _amazing_ if you could fix this in ironic by adding a "last cleaned time" or something btw). The only way to deal with this is to write some change to the BareMetalHost Status when you start the cleaning.

One option is to leave the "Registering" state as soon as the cleaning has started, rather than wait for it to finish. A problem with that is that if it fails then the errors are reported in the next state rather than in the Registering state - also you have the problem of how to retry. The other option is to add some "hasBeenReset" flag to the status. A problem with that is that the user is also going to be looking at the status and it's not going to make much sense to them.

Probably the least-worst option would be to start the cleaning in the Registering state, then immediately move on to a new Resetting state that will handle errors and retries.